File: metaapp/src/main/java/com/sourcepointmeta/metaapp/viewmodel/PropertyListViewModel.java
package com.sourcepointmeta.metaapp.viewmodel;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.ViewModel;

import com.sourcepointmeta.metaapp.database.entity.Property;
import com.sourcepointmeta.metaapp.repository.PropertyListRepository;

import java.util.List;

public class PropertyListViewModel extends ViewModel {

    private final PropertyListRepository mPropertyListRepository;

    public PropertyListViewModel(PropertyListRepository repository) {
        mPropertyListRepository = repository;
    }

    public LiveData<List<Property>> getPropertyListLiveData() {
        return mPropertyListRepository.showPropertyList();
    }


    public MutableLiveData<Integer> deleteProperty(Property property){
       return mPropertyListRepository.deleteProperty(property);
    }

}

File: samples/metaapp/src/main/java/com/sourcepointmeta/metaapp/ui/PropertyListViewModel.kt
package com.sourcepointmeta.metaapp.ui

import androidx.lifecycle.* // ktlint-disable
import com.sourcepointmeta.metaapp.core.fold
import com.sourcepointmeta.metaapp.data.localdatasource.LocalDataSource
import com.sourcepointmeta.metaapp.data.localdatasource.Property
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlin.coroutines.CoroutineContext

internal class PropertyListViewModel(
    private val dataSource: LocalDataSource,
    private val workerDispatcher: CoroutineContext = Dispatchers.IO
) : ViewModel() {

    private val mutableLiveData by lazy { MutableLiveData<BaseState>() }
    val liveData: LiveData<BaseState> get() = mutableLiveData

    fun fetchPropertyList() {
        viewModelScope.launch {
            val either = withContext(workerDispatcher) { dataSource.fetchProperties() }
            either.fold(
                { /* handle the exception */ },
                {
                    mutableLiveData.value = BaseState.StateSuccess(it)
                }
            )
        }
    }

    fun updateProperty(property: Property) {
        viewModelScope.launch(workerDispatcher) { dataSource.storeOrUpdateProperty(property) }
    }

    fun deleteProperty(property: Property) {
        viewModelScope.launch(workerDispatcher) { dataSource.deleteByPropertyName(property.propertyName) }
        fetchPropertyList()
    }

    fun deleteProperty(propertyName: String) {
        viewModelScope.launch(workerDispatcher) { dataSource.deleteByPropertyName(propertyName) }
        fetchPropertyList()
    }

    fun clearDB() {
        viewModelScope.launch(workerDispatcher) {
            dataSource.deleteAll()
        }
    }
}
